tests/gitsvntests.py

```diff
@@ -0,0 +1,241 @@
+        with open(os.path.join(self.pkgdir, '_servicedata'), 'w') as sd:
+            sd.write(textwrap.dedent("""\
+              <servicedata>
+                <service name="tar_scm">
+                  <param name="url">%s</param>
+                </service>
+              </servicedata>
+            """ % self.fixtures.repo_url))
+        self.tar_scm_std('--changesgenerate', 'enable')
+        self._check_servicedata()
+
+    def _write_changes_file(self):
+        contents = textwrap.dedent("""\
+          -------------------------------------------------------------------
+          Fri Oct  3 00:17:50 BST 2014 - %s
+
+          - 2
+
+          -------------------------------------------------------------------
+          Thu Sep 18 10:27:14 BST 2014 - %s
+
+          - 1
+        """ % (self.fixtures.user_email, self.fixtures.user_email))
+        with open(os.path.join(self.pkgdir, 'pkg.changes'), 'w') as f:
+            f.write(contents)
+        return contents
+
+    def test_changesgenerate_no_change_or_changes_file(self):
+        self._write_servicedata(2)
+        self.tar_scm_std('--changesgenerate', 'enable')
+        self._check_servicedata()
+
+    def test_changesgenerate_no_change_same_changes_file(self):
+        self._write_servicedata(2)
+        orig_changes = self._write_changes_file()
+        self.tar_scm_std('--changesgenerate', 'enable')
+        self._check_servicedata()
+
+    def test_changesgenerate_new_commit_no_changes_file(self):
+        self._write_servicedata(2)
+        self.fixtures.create_commits(1)
+        self.tar_scm_std('--changesgenerate', 'enable')
+        self._check_servicedata(revision=3)
+
+    def _new_change_entry_regexp(self, author, changes):
+        return textwrap.dedent("""\
+          ^-------------------------------------------------------------------
+          \w{3} \w{3} [ \d]\d \d\d:\d\d:\d\d [A-Z]{3} 20\d\d - %s
+
+          %s
+          """) % (author, changes)
+
+    def _check_changes(self, orig_changes, expected_changes_regexp):
+        new_changes_file = os.path.join(self.outdir, 'pkg.changes')
+        self.assertTrue(os.path.exists(new_changes_file))
+        with open(new_changes_file) as f:
+            new_changes = f.read()
+            self.assertNotEqual(orig_changes, new_changes)
+            print new_changes
+            expected_changes_regexp += "(.*)"
+            self.assertRegexpMatches(new_changes, expected_changes_regexp)
+            m = re.match(expected_changes_regexp, new_changes, re.DOTALL)
+            self.assertEqual(m.group(1), orig_changes)
+
+    def test_changesgenerate_new_commit_and_changes_file(self):
+        self._test_changesgenerate_new_commit_and_changes_file(
+            self.fixtures.user_email)
+
+    def test_changesgenerate_new_commit_and_changes_file_default_author(self):
+        self._test_changesgenerate_new_commit_and_changes_file()
+
+    def _write_servicedata(self, rev):
+        if self.scm == 'git':
```

Comment:
Thanks - having this method in the shared class is definitely an improvement, but it would be cleaner to have the bits specific to svn or git outside the shared class.  For example,

    class GitTests(GitHgTests):
        ...
        def changesrevision(rev):
            return self.sha1s('tag%d' % rev)
        ...

    class SvnTests(CommonTests):
        ...
        def changesrevision(rev):
            return rev
        ...